### ISSUE TITLE
koboldcpp token usage

### ISSUE THREAD

**Issue by Enferlain** [Cohee1207/SillyTavern#45]
If I let's say have tokens set at 200, why does it fill the entire 200 tokens instead of cutting the reply when it finishes writing an answer? There is an option for single line mode in both koboldcpp and sillytavern, but it doesn't stop generation like it does in ooba webui, it only stops them from appearing in the ui, so the wait times are long even if the returned reply is like 10 words. And if I set the token count too low, then the replies will have no potential to be longer

Is there a way to stop it from having a conversation with itself to fill the entire token limit? 

Also, what is the processing prompt BLAS thing that takes 2-3 minutes to go through frequently? And is there a way to enable text streaming?

Thanks!

**Comment by Enferlain** [comment]
Thanks for the reply. 

In the meanwhile I found out that the ai conversing with itself behavior was due to the gpt4x model I was using. It showed the same exact behavior in ooba webui and I've seen others talk about it before. Likely just related to how the models were trained, so that one is easy to fix by just changing models.

The single-line mode function works, as in it won't show the generated lines after the first one, but yeah generation continues even if it's enabled, and same with the koboldcpp equivalent setting so I will be asking over there as well.

As for the processing prompt, for me it kept reprocessing about 1800 tokens worth around every 4-5 generations, even when I wasn't messing with the configuration, that's why I wanted to ask about it.

**Comment by Enferlain** [comment]
That's really huge, thanks for all the information. I did have some stuff that inserted context into the start, so that explains the constant prompt processing.

I tried multigen briefly, but I didn't use it enough to explore if it helps with my problem, so I'll give it a go again, it sounds really nice in the way you put it. 

The main issue really is the fact that there is no way to stop generation from doing weird things like filling the tokens with random conversation. I did test a token size of like 8 in koboldcpp directly, with auto adjust limits enabled, and it somewhat did what you're describing I think, but it still attempted to insert lines in place of other participants (me) instead of just replying, and for some reason this happens in every single model, so it's not just related to the ones that were trained in ### human ### assistant format.

That's why I think it would be nice if something similar to ooba webui "stop generation at new line" could be implemented, since it would eliminate the root issue and also speed up conversations. Longer replies would still remain long, without having to potentially wait for too long only to get nothing worthwhile returned.

**Comment by Necroxix** [comment]
I did a bit more thinking about the reprocessing problem. I'm realizing that once the context size exceeds the limit, tavern is going to be cutting early parts of the prompt out to fit the newest stuff anyways so it's going to keep changing and making it process. I think inevitably we have to wait until someone can come up with a clever solution on the back end :(

**Comment by Enferlain** [comment]
Downloaded the entire dev branch, have single line on with default 180 token size and it still goes through the entire 180 tokens, and terminal is 
Character: 
You: 
Character: 
So it doesn't seem like it stops generating after new line. I think it's a koboldcpp problem first, rather than tavern. Already asked over there as well, so hopefully it will get implemented sooner or later.

As for the processing of the prompt, I didn't notice anything different. I tried with multigen as well and it will process the entire full context twice if I have it enabled. Hopefully Necroxix will be able to test as well and see how it goes.

**Comment by Necroxix** [comment]
I finally had a little time to try things out a little today. The ooba text gen mode does look like it cuts off now at \nUsername as you said without multi-gen and even seemed faster. 

Some quick testing with multi-gen though, it seems to keep going. Tried with both ooba mode and koboldcpp mode. I can see in the first 50 token generations that it finished the character's dialog and continuing the dialog with Username: stuffs, but it continues to iterate through the additional 30 token generations needlessly.

I'll play around with it more as time allows.

**Comment by Enferlain** [comment]
Yeah ooba api works with stopping, but for me ooba is way slower which is why I used koboldcpp in the first place. It's like running CPU mode on GPU models. 3 minutes for 10 words type experience.

"Quick edit: It does stop generating after the first 50 token pass when I ticked pygmalion formatting, which changed the username to "You:" in the console."

I'll try this

**Comment by Necroxix** [comment]
@Enferlain Here's the best info I have for you as end users at the moment based on my findings with current versions of SillyTavern and Koboldcpp when using LLaMA type models. Hopefully this assists others.

1) Koboldcpp's end of things. The generation time problem. Yesterday-ish they added two parameters you can use: --useclblast 0 0 and --smartcontext. Give those a shot. the clblast one does some gpu stuff. The smartcontext tries to do some optimizations for generation time on the prompt once you've hit the full context size which is the real issue. It's still a hackjob but it may help reduce the generation time. Keep tabs on their project, I think they're still trying to find better solutions.

2) SillyTavern's end of things: Enabling multigen + pygmalion formatting. This helps reduce the generation time itself in instances where the response happens to be less than your "response length (tokens)" you set. So if you had set 200 tokens for responses,  it doesn't generate for the entire 200 length when the response is just "Waifu: How are you? \nYou: blahblah", it will stop after 50 taking less time to generate needlessly. It does still show the You: and stuff in the console, it's just generating less of it because of the smaller token length. The benefit is simply less generation time if the response is shorter. It's basically just an end-user hack that happens to work because multi-gen exists.

Thoughts for @Cohee1207: This may not have been the intended use case for multigen but perhaps let the user customize the 50/30 token values used in multigen. If those values were allowed to be set lower manually, it may allow generation times to be reduced even further for this kind of use case without messing it up for it's intended use case otherwise.

Not sure if your code is checking for "You:" across the combined generations, or just per iteration string. IE if it was only set for 8 tokens and the first generation cuts off mid way at like "\nYo" and the second generation resumes with "u: blahblah", will it stop after the second generation because it checks across the combined 16 token generation or completely miss it and keep going into the next 8.

I think you mentioned you were already planning on adding customizable stop strings so I'd like mention one thing I noticed in my testing yesterday. I tested editing the script.js multigen code that checks for 'You:' and '<|endoftext|>' and added more && for my own use case testing by adding my own name 'Necro:' allowing me to turn off the pygma formatting and it worked for multigen stopping but the GUI didn't trim that out. So if you do make it possible for it to be a list of custom stop strings, it may be needed to apply that list for both multigen and however the GUI trims the output as well. 

I was using a deterministic generation setting and in my model hit unique moments where it generates the first line properly then goes into spitting out non-dialog stuff, like starting the next line with '<END>' and 'end]' so I added those to check as well and it allowed mutigen to stop though the GUI still showed that stuff. So having that custom list I can add these edge cases to as a user will be really useful to tame models not intended for chatting into functioning.

Sorry if this is all long winded. I did a new ooba install directory to test the newest version today and finally saw the streaming feature in SillyTavern that I guess you added. Only tried it for a bit but it's pretty cool. If you could find it in your heart to include the generation time in the GUI somewhere I would be grateful! I don't want to derail this from koboldcpp so I'll just leave it at this.

**Comment by Necroxix** [comment]
Sure! Testing using koboldcpp with the gpt4-x-alpaca-13b-native-ggml-model using multigen at default 50x30 batch settings and generation settings set to 400 tokens. **It appears to be working in all 3 modes and stopping at either "You:" or my "Name:" without trying to continue to 400 tokens and taking forever to finish generating**.

Disabled for all models: Console shows my name and stopped after that iteration
![aoMoGqs](https://user-images.githubusercontent.com/130382665/232250665-535f48ad-b0aa-4ef2-af69-233fda1664c5.png)

Automatic: Same as disabled. It saw my name and stopped after that iteration
![sxZGD2x](https://user-images.githubusercontent.com/130382665/232250722-981c6af2-fa6b-403a-aa3d-27de42fa03c7.png)

Enabled: Saw the "You:" and stopped after that iteration
![DQykp0f](https://user-images.githubusercontent.com/130382665/232250774-8d3fe658-c425-4c66-a7eb-76f0fb828a2c.png)

-

One anomaly worth sharing: since the model isn't trained specifically for chat it can add weird stuff after.
![xltBY64](https://user-images.githubusercontent.com/130382665/232250888-eadf8ccd-fe28-497b-be8f-b221be74092b.png)

That's the type of edge case where I would rather have control over it as a user and manually enter <END> to a list of stop strings, rather than having it hard coded. 

But that aside, the username/you stopping does seem to be working exactly as intended.

**Comment by Enferlain** [comment]
It's actually really good now. Like compared to when I first tried and ooba webui with ggml models it's way faster. On ooba it's like using cpu on gpu models, and with koboldcpp now having smartcontext and some other optimizations this works nicely.
